### ISSUE TITLE
false-positive for CA2013

### ISSUE THREAD

**Issue by collinstevens** [dotnet/roslyn-analyzers#6149]
### Analyzer

**Diagnostic ID**: [CA2013](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2013): `Do not use ReferenceEquals with value types`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.303](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)

### Describe the bug

The CA2013 analyzer is showing a warning for the `ReferenceEquals` in the below code snippet with the following text `Do not pass an argument with value type 'int?' to 'ReferenceEquals'. Due to value boxing, this call to 'ReferenceEquals' will always return 'false'.`.

Nevertheless, the the call to `ReferenceEquals` does not return `false` as the analyzer would indicate, but instead returns `true`.

The following xUnit test passes.

```
[Fact]
public void ShouldReturnTrueForNullIntegers()
{
    // Arrange
    int? first = null;
    int? second = null;

    // Act
    var result = ReferenceEquals(first, second);

    // Assert
    Assert.True(result);
}
```

### Steps To Reproduce

Execute the provided xUnit test.

### Expected behavior

The analyzer should not warn for Nullable<T>.

### Actual behavior

The analyzer should warns for Nullable<T>.

### Additional context

N/A

**Comment by svick** [comment]
The error message is inaccurate, and should be improved. But I think the warning should still be shown here, since `ReferenceEquals` does not make sense for `Nullable<T>` and you should use regular equality instead.

**Comment by Youssef1313** [comment]
This involves updating three resources:

https://github.com/dotnet/roslyn-analyzers/blob/a26d1a203d0116aa769bce0fd3a917cc70d88a66/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx#L933-L941

My suggestion:

```xml
 <data name="DoNotUseReferenceEqualsWithValueTypesDescription" xml:space="preserve"> 
   <value>Value type typed arguments are uniquely boxed for each call to this method, therefore the result can be unexpected or incorrect.</value> 
 </data> 
 <data name="DoNotUseReferenceEqualsWithValueTypesComparerMessage" xml:space="preserve"> 
   <value>Do not pass an argument with value type '{0}' to the 'Equals' method on 'ReferenceEqualityComparer'. Due to value boxing, this call to 'Equals' can return unexpected or incorrect result.</value> 
 </data> 
 <data name="DoNotUseReferenceEqualsWithValueTypesMethodMessage" xml:space="preserve"> 
   <value>Do not pass an argument with value type '{0}' to 'ReferenceEquals'. Due to value boxing, this call to 'ReferenceEquals' can return unexpected or incorrect result.</value> 
 </data> 
```
